File: final_assignment_front/lib/features/model/login_log.dart
class LoginLog {
  int? logId;
  String? username;
  String? loginIpAddress;
  DateTime? loginTime;
  String? loginResult;
  String? browserType;
  String? osVersion;
  String? remarks;
  String? idempotencyKey;

  LoginLog({
    this.logId,
    this.username,
    this.loginIpAddress,
    this.loginTime,
    this.loginResult,
    this.browserType,
    this.osVersion,
    this.remarks,
    this.idempotencyKey,
  });

  @override
  String toString() {
    return 'LoginLog[logId=$logId, username=$username, loginIpAddress=$loginIpAddress, loginTime=$loginTime, loginResult=$loginResult, browserType=$browserType, osVersion=$osVersion, remarks=$remarks, idempotencyKey=$idempotencyKey]';
  }

  factory LoginLog.fromJson(Map<String, dynamic> json) {
    return LoginLog(
      logId: json['logId'] as int?,
      username: json['username'] as String?,
      loginIpAddress: json['loginIpAddress'] as String?,
      loginTime: json['loginTime'] != null ? DateTime.parse(json['loginTime'] as String) : null,
      loginResult: json['loginResult'] as String?,
      browserType: json['browserType'] as String?,
      osVersion: json['osVersion'] as String?,
      remarks: json['remarks'] as String?,
      idempotencyKey: json['idempotencyKey'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> json = {};
    if (logId != null) json['logId'] = logId;
    if (username != null) json['username'] = username;
    if (loginIpAddress != null) json['loginIpAddress'] = loginIpAddress;
    if (loginTime != null) json['loginTime'] = loginTime!.toIso8601String();
    if (loginResult != null) json['loginResult'] = loginResult;
    if (browserType != null) json['browserType'] = browserType;
    if (osVersion != null) json['osVersion'] = osVersion;
    if (remarks != null) json['remarks'] = remarks;
    if (idempotencyKey != null) json['idempotencyKey'] = idempotencyKey;
    return json;
  }

  static List<LoginLog> listFromJson(List<dynamic> json) {
    return json.map((value) => LoginLog.fromJson(value as Map<String, dynamic>)).toList();
  }

  static Map<String, LoginLog> mapFromJson(Map<String, dynamic> json) {
    var map = <String, LoginLog>{};
    if (json.isNotEmpty) {
      json.forEach((key, value) => map[key] = LoginLog.fromJson(value));
    }
    return map;
  }

  static Map<String, List<LoginLog>> mapListFromJson(Map<String, dynamic> json) {
    var map = <String, List<LoginLog>>{};
    if (json.isNotEmpty) {
      json.forEach((key, value) => map[key] = LoginLog.listFromJson(value));
    }
    return map;
  }
}

File: final_assignment_backend_go/project/internal/domain/login_log.go
package domain

import (
	"time"

	"gorm.io/gorm"
)

// LoginLog maps traffic.audit_login_log and emits camelCase JSON.
type LoginLog struct {
	LogID          int            `gorm:"column:log_id;primaryKey;autoIncrement" json:"logId"`
	Username       string         `gorm:"column:username" json:"username"`
	LoginTime      time.Time      `gorm:"column:login_time" json:"loginTime"`
	LogoutTime     *time.Time     `gorm:"column:logout_time" json:"logoutTime"`
	LoginResult    string         `gorm:"column:login_result" json:"loginResult"`
	FailureReason  string         `gorm:"column:failure_reason" json:"failureReason"`
	LoginIP        string         `gorm:"column:login_ip" json:"loginIp"`
	LoginIPAddress string         `gorm:"-" json:"loginIpAddress"`
	LoginLocation  string         `gorm:"column:login_location" json:"loginLocation"`
	BrowserType    string         `gorm:"column:browser_type" json:"browserType"`
	BrowserVersion string         `gorm:"column:browser_version" json:"browserVersion"`
	OSType         string         `gorm:"column:os_type" json:"osType"`
	OSVersion      string         `gorm:"column:os_version" json:"osVersion"`
	DeviceType     string         `gorm:"column:device_type" json:"deviceType"`
	UserAgent      string         `gorm:"column:user_agent" json:"userAgent"`
	SessionID      string         `gorm:"column:session_id" json:"sessionId"`
	CreatedAt      *time.Time     `gorm:"column:created_at" json:"createdAt"`
	Remarks        string         `gorm:"column:remarks" json:"remarks"`
	DeletedAt      gorm.DeletedAt `gorm:"column:deleted_at;index" json:"-"`
}

func (LoginLog) TableName() string {
	return "audit_login_log"
}

func (l *LoginLog) AfterFind(tx *gorm.DB) error {
	l.LoginIPAddress = l.LoginIP
	return nil
}

func (l *LoginLog) BeforeSave(tx *gorm.DB) error {
	if l.LoginIP == "" && l.LoginIPAddress != "" {
		l.LoginIP = l.LoginIPAddress
	}
	return nil
}
